File: +ndi/+daq/+reader/+mfdaq/spikegadgets.m
% NDI_DAQREADER_MFDAQ_SPIKEGADGETS - Device driver for SpikeGadgets .rec video file format
%
% This class reads data from video files .rec that spikegadgets use
%
% Spike Gadgets: http://spikegadgets.com/
%
%

classdef spikegadgets < ndi.daq.reader.mfdaq

	properties
	end

	methods
		function obj = spikegadgets(varargin)
			% NDI_DAQSYSTEM_MFDAQ_SPIKEGADGETS - Create a new NDI_DEVICE_MFDAQ_SPIKEGADGETS object
			%
			%  D = NDI_DAQSYSTEM_MFDAQ_SPIKEGADGETS(NAME,THEFILENAVIGATOR)
			%
			%  Creates a new NDI_DAQSYSTEM_MFDAQ_SPIKEGADGETS object with name NAME and associated
			%  filenavigator THEFILENAVIGATOR.
			%
			%
				obj = obj@ndi.daq.reader.mfdaq(varargin{:});
		end

		function channels = getchannelsepoch(ndi_daqreader_mfdaq_spikegadgets_obj, epochfiles)
			% GETCHANNELSEPOCH - GET THE CHANNELS AVAILABLE FROM .REC FILE HEADER
			%
			% CHANNELS = GETCHANNELSEPOCH(NDI_DAQREADER_MFDAQ_SPIKEGADGETS_OBJ)
			%
			% CHANNELS is a STRUCT
        
				filename = ndi_daqreader_mfdaq_spikegadgets_obj.filenamefromepochfiles(epochfiles); 
				fileconfig = [];
				[fileconfig, channels] = read_SpikeGadgets_config(filename);

				for k=1:length(channels)
					number = 0;
					name = '';

					%Auxiliary
					if strcmp(channels(k).name(1),'A')
						%Input
						if strcmp(channels(k).name(2),'i')
							channels(k).type = 'auxiliary';
							number = sscanf(channels(k).name, 'Ain%d'); 
							name = strcat('axn',num2str(number));
							channels(k).number = number;
							channels(k).time_channel = 1;
								%Output
						else
							channels(k).type = 'auxiliary';
							number = sscanf(channels(k).name, 'Aout%d'); 
							name = strcat('axo',num2str(number));
							channels(k).number = number;
							channels(k).time_channel = 1;
						end

					%Digital
					elseif strcmp(channels(k).name(1),'D')
						if strcmp(channels(k).name(2),'i') % Input
							channels(k).type = 'digital_in';
							number = sscanf(channels(k).name, 'Din%d'); 
							name = strcat('di',num2str(number));
							channels(k).number = number;
							channels(k).time_channel = 1;
						else %Output
							channels(k).type = 'digital_out';
							number = sscanf(channels(k).name, 'Dout%d');
							name = strcat('do',num2str(number));
							channels(k).number = number;
							channels(k).time_channel = 1;
						end
					else	%MCU (digital inputs)
						channels(k).type = 'digital_in';
						number = sscanf(channels(k).name, 'MCU_Din%d');
						number = number + 32; % +32 from previous non MCU inputs
						name = strcat('di',num2str(number));
						channels(k).number = number;
						channels(k).time_channel = 1;
					end
					channels(k).name = name;
				end

				%Adds all nTrodes to the list
				for i=1:length(fileconfig.nTrodes)
					for j=1:4 %argument for 4 channels, variable could be used later to deal with this in a more general way
						channelNumber = fileconfig.nTrodes(i).channelInfo(j).packetLocation;
						channels(end+1).name = strcat('ai',num2str(channelNumber+1));
						channels(end).type = 'analog_in';
						channels(end).number = channelNumber+1;
						channels(end).time_channel = 1;
					end
				end

				channels = struct2table(channels);
				channels = sortrows(channels,{'type','number'});
				channels = table2struct(channels);

				remove = {'startbyte','bit','number'};

				channels = rmfield(channels, remove);
		end

		function channels = getchannelsepochdetailed(ndi_daqreader_mfdaq_spikegadgets_obj, epochfiles)
			% GETCHANNELSDETAILED - GET THE CHANNELS AVAILABLE FROM .REC FILE HEADER WITH EXTRA DETAILS
			%
			% CHANNELS = GETCHANNELSEPOCHDETAILED(NDI_DAQREADER_MFDAQ_SPIKEGADGETS_OBJ)
			%
			% CHANNELS is a STRUCT

				filename = ndi_daqreader_mfdaq_spikegadgets_obj.filenamefromepochfiles(epochfiles); 
				fileconfig = [];

				[fileconfig, channels] = read_SpikeGadgets_config(filename);

				for k=1:length(channels)
					number = 0;
					name = '';
					%
					%Auxiliary
					if strcmp(channels(k).name(1),'A')
						%Input
						if strcmp(channels(k).name(2),'i')
						    channels(k).type = 'auxiliary';
						    number = sscanf(channels(k).name, 'Ain%d'); %number = channels(k).name(4:end);
						    name = strcat('axn',num2str(number));
						    channels(k).number = number;
						%Output
						else
						    channels(k).type = 'auxiliary';
						    number = sscanf(channels(k).name, 'Aout%d'); %number = channels(k).name(5:end);
						    name = strcat('axo',num2str(number));
						    channels(k).number = number;
						end

					%Digital
					elseif strcmp(channels(k).name(1),'D')
						%Input
						if strcmp(channels(k).name(2),'i')
						    channels(k).type = 'digital_in';
						    number = sscanf(channels(k).name, 'Din%d'); %number = channels(k).name(4:end);
						    name = strcat('di',num2str(number));
						    channels(k).number = number;
						%Output
						else
						    channels(k).type = 'digital_out';
						    number = sscanf(channels(k).name, 'Dout%d'); %number = channels(k).name(5:end);
						    name = strcat('do',num2str(number));
						    channels(k).number = number;
						end
					%MCU (digital inputs)
					else
						channels(k).type = 'digital_in';
						number = sscanf(channels(k).name, 'MCU_Din%d'); %str2num(channels(k).name(8:end));
						number = number + 32; % +32 from previous non MCU inputs
						name = strcat('di',num2str(number));
						channels(k).number = number;
					end

					channels(k).name = name;
				end


				%Adds all nTrodes to the list
				for i=1:length(fileconfig.nTrodes)
					for j=1:4
						channelNumber = fileconfig.nTrodes(i).channelInfo(j).packetLocation;
						channels(end+1).name = strcat('ai',num2str(channelNumber+1));
						channels(end).type = 'analog_in';
						channels(end).number = channelNumber;
					end
				end

				channels = struct2table(channels);
				channels = sortrows(channels,{'type','number'});
				channels = table2struct(channels);
		end

		function sr = samplerate(ndi_daqreader_mfdaq_spikegadgets_obj, epochfiles, channeltype, channel)
			% SAMPLERATE - GET THE SAMPLE RATE FOR SPECIFIC EPOCH AND CHANNEL
			%
			% SR = SAMPLERATE(DEV, EPOCHFILES, CHANNELTYPE, CHANNEL)
			%
			% SR is the list of sample rate from specified channels
			%
			% CHANNELTYPE and CHANNEL not used in this case since it is the
			% same for all channels in this device

				filename = ndi_daqreader_mfdaq_spikegadgets_obj.filenamefromepochfiles(epochfiles); 

				fileconfig = read_SpikeGadgets_config(filename);

				%Sampling rate is the same for all channels in Spike Gadgets
				%device so it is returned by checking the file configuration
				sr = str2num(fileconfig.samplingRate);
		end

		function t0t1 = t0_t1(ndi_daqreader_mfdaq_spikegadgets_obj, epochfiles)
			% EPOCHCLOCK - return the t0_t1 (beginning and end) epoch times for an epoch
			%
			% T0T1 = T0_T1(NDI_EPOCHSET_OBJ, EPOCHFILES)
			%
			% Return the beginning (t0) and end (t1) times of the epoch EPOCH_NUMBER
			% in the same units as the ndi.time.clocktype objects returned by EPOCHCLOCK.
			%
			% The abstract class always returns {[NaN NaN]}.
			%
			% See also: ndi.time.clocktype, EPOCHCLOCK
			%
				filename = ndi_daqreader_mfdaq_spikegadgets_obj.filenamefromepochfiles(epochfiles); 

				[fileconfig, ~] = read_SpikeGadgets_config(filename);

				headerSizeBytes = str2num(fileconfig.headerSize) * 2; % int16 = 2 bytes
				channelSizeBytes = str2num(fileconfig.numChannels) * 2; % int16 = 2 bytes
				blockSizeBytes = headerSizeBytes + 2 + channelSizeBytes;

				s = dir(filename);

				bytes_present = s.bytes;

				bytes_per_block = blockSizeBytes;

				num_data_blocks = (bytes_present - headerSizeBytes) / bytes_per_block;

				total_samples = num_data_blocks;
				total_time = (total_samples - 1) / str2num(fileconfig.samplingRate); % in seconds

				t0 = 0;
				t1 = total_time;

				t0t1 = {[t0 t1]};
		end % t0t1

		function epochprobemap = getepochprobemap(ndi_daqreader_mfdaq_spikegadgets_obj, epochmapfilename, epochfiles)
		        % GETEPOCHPROBEMAP returns struct with probe information
		        % name, reference, n-trode, channels
		        %
				filename = ndi_daqreader_mfdaq_spikegadgets_obj.filenamefromepochfiles(epochfiles);
				fileconfig = read_SpikeGadgets_config(filename);
				nTrodes = fileconfig.nTrodes;
				%List where epochprobemap objects will be stored
				epochprobemap = [];

				for i=1:length(nTrodes)
					name = strcat('Tetrode', nTrodes(i).id);
					reference = 1;
					type = 'n-trode';
					channels = [];

					for j=1:length(nTrodes(i).channelInfo) %number of channels per nTrode
						%Array with channels of trode
						channels = [channels nTrodes(i).channelInfo(j).packetLocation + 1];
					end
					%Object that deals with channels
					devicestringobject = ndi.daq.daqsystemstring('SpikeGadgets',{'ai','ai','ai','ai'}, channels);
					devicestringstring = devicestringobject.devicestring();
					% FIX: we need some way of specifying the subject, which is not in the file to my knowledge (although maybe it is)
					obj = ndi.epoch.epochprobemap_daqsystem(name,reference,type,devicestringstring,'anteater52@nosuchlab.org');
					%Append each newly made object to end of list
					epochprobemap = [epochprobemap obj];
				end
        	end

		function data = readchannels_epochsamples(ndi_daqreader_mfdaq_spikegadgets_obj, channeltype, channels, epochfiles, s0, s1)
			% FUNCTION READ_CHANNELS - read the data based on specified channels
			%
			% DATA = READ_CHANNELS(MYDEV, CHANNELTYPE, CHANNEL, EPOCHFILES ,S0, S1)
			%
			% CHANNELTYPE is the type of channel to read
			% 'digital_in', 'digital_out', 'analog_in', 'analog_out' or 'auxiliary'
			%
			% CHANNEL is a vector of the channel numbers to
			% read beginning from 1 if 'etrodeftrode' is channeltype,
			% if channeltype is 'analog_in' channel is an array with the
			% string names of analog channels 'Ain1'through 8
			%
			% EPOCH is set of files in the epoch
			%
			% DATA is the channel data (each column contains data from an indvidual channel)
			%
				filename = ndi_daqreader_mfdaq_spikegadgets_obj.filenamefromepochfiles(epochfiles); 

				header = read_SpikeGadgets_config(filename);

				sr = ndi_daqreader_mfdaq_spikegadgets_obj.samplerate(epochfiles,channeltype,channels);

				detailedchannels = ndi_daqreader_mfdaq_spikegadgets_obj.getchannelsepochdetailed(epochfiles);

				byteandbit = [];
                
                data = [];
                
				%read_SpikeGadgets_trodeChannels(filename,NumChannels, channels,samplingRate,headerSize, configExists)
				%reading from channel 1 in list returned
				%Reads nTrodes
				%WARNING channeltype hard coded, ask Steve
				channeltype
				if (strcmp(ndi.daq.system.mfdaq.mfdaq_type(channeltype{1}),'analog_in') || strcmp(ndi.daq.system.mfdaq.mfdaq_type(channeltype{1}), 'analog_out'))
					data = read_SpikeGadgets_trodeChannels(filename,header.numChannels,channels-1,sr, header.headerSize,s0,s1);

				elseif (strcmp(channeltype,'auxiliary') || strcmp(channeltype,'aux')) %Reads analog inputs
					%for every channel in device
					for i=1:length(detailedchannels)
						%based on every channel to read
						for j=1:length(channels)
							%check if channel number and channeltype match
							if (strcmp(detailedchannels(i).type,'auxiliary') && detailedchannels(i).number == channels(j))
								%add startbyte to list of channels to read
								byteandbit(end+1) = str2num(detailedchannels(i).startbyte);
							end
						end
					end
					data = read_SpikeGadgets_analogChannels(filename,header.numChannels,byteandbit,sr,header.headerSize,s0,s1);

				elseif (strcmp(channeltype,'digital_in') || strcmp(channeltype, 'digital_out')), %Reads digital inputs
					%for every channel in device
					for i=1:length(detailedchannels)
						%based on every channel to read
						for j=1:length(channels)
							%check if channel number and channeltype match
							if (strcmp(detailedchannels(i).type,channeltype) && detailedchannels(i).number == channels(j))
								%add startbyte to list of channels to read
								byteandbit(end+1,1) = str2num(detailedchannels(i).startbyte);
								byteandbit(end,2) = str2num(detailedchannels(i).bit) + 1;
							end
						end
					end

					data = read_SpikeGadgets_digitalChannels(filename,header.numChannels,byteandbit,sr,header.headerSize,s0,s1);
					data = data';
				else

				end
		end % readchannels_epochsamples

		function filename = filenamefromepochfiles(ndi_daqreader_mfdaq_spikegadgets_obj, filename)
				s1 = ['.*\.rec\>']; % equivalent of *.ext on the command line
				[tf, matchstring, substring] = vlt.string.strcmp_substitution(s1,filename,'UseSubstituteString',0);
				index = find(tf);
				if numel(index)> 1,
					error(['Need only 1 .rec file per epoch.']);
				elseif numel(index)==0,
					error(['Need 1 .rec file per epoch.']);
				else,
					filename = filename{index};
				end
                end % filenamefromepoch

    end % methods
end % classdef

File: ndi-python/ndi/daq/readers/mfdaq/spikegadgets.py
"""
NDI DAQ Reader MFDAQ SpikeGadgets - Device driver for SpikeGadgets .rec video file format.

This class reads data from video files .rec that SpikeGadgets uses.

SpikeGadgets: http://spikegadgets.com/
"""

from typing import List, Tuple, Dict, Any, Optional
import re
import os
import numpy as np
from ..mfdaq import MFDAQReader


class SpikeGadgets(MFDAQReader):
    """
    SpikeGadgets reader for SpikeGadgets .rec file format.

    This class reads data from SpikeGadgets .rec files, supporting
    analog inputs, auxiliary inputs, digital inputs/outputs, and nTrodes.

    Examples:
        >>> from ndi.daq.reader.mfdaq import SpikeGadgets
        >>> reader = SpikeGadgets()
        >>> channels = reader.getchannelsepoch(epochfiles)
    """

    def __init__(self, *args):
        """
        Create a new SpikeGadgets reader.

        Args:
            *args: Arguments passed to MFDAQReader constructor
        """
        super().__init__(*args)

    def getchannelsepoch(self, epochfiles: List[str]) -> List[Dict[str, Any]]:
        """
        Get the channels available from .rec file header.

        Args:
            epochfiles: List of file paths for this epoch

        Returns:
            List of channel dictionaries with fields:
            - 'name': The name of the channel (e.g., 'ai1', 'di1', 'ax1')
            - 'type': The type of data stored in the channel
        """
        filename = self.filenamefromepochfiles(epochfiles)
        fileconfig, channels_raw = self._read_spikegadgets_config(filename)

        channels = []

        # Process auxiliary and digital channels
        for ch_raw in channels_raw:
            ch_name = ch_raw.get('name', '')
            number = 0
            name = ''
            ch_type = ''
            time_channel = 1

            # Auxiliary channels
            if ch_name.startswith('A'):
                if len(ch_name) > 1 and ch_name[1] == 'i':
                    # Ain# - Auxiliary input
                    ch_type = 'auxiliary'
                    number = self._parse_channel_number(ch_name, 'Ain')
                    name = f'axn{number}'
                elif len(ch_name) > 1 and ch_name[1] == 'o':
                    # Aout# - Auxiliary output
                    ch_type = 'auxiliary'
                    number = self._parse_channel_number(ch_name, 'Aout')
                    name = f'axo{number}'

            # Digital channels
            elif ch_name.startswith('D'):
                if len(ch_name) > 1 and ch_name[1] == 'i':
                    # Din# - Digital input
                    ch_type = 'digital_in'
                    number = self._parse_channel_number(ch_name, 'Din')
                    name = f'di{number}'
                elif len(ch_name) > 1 and ch_name[1] == 'o':
                    # Dout# - Digital output
                    ch_type = 'digital_out'
                    number = self._parse_channel_number(ch_name, 'Dout')
                    name = f'do{number}'

            # MCU digital inputs
            elif ch_name.startswith('MCU_Din'):
                ch_type = 'digital_in'
                number = self._parse_channel_number(ch_name, 'MCU_Din')
                number += 32  # Offset from non-MCU inputs
                name = f'di{number}'

            if name:
                channels.append({
                    'name': name,
                    'type': ch_type,
                    'time_channel': time_channel
                })

        # Add nTrode channels (analog inputs)
        if fileconfig and 'nTrodes' in fileconfig:
            for ntrode in fileconfig['nTrodes']:
                if 'channelInfo' in ntrode:
                    for ch_info in ntrode['channelInfo']:
                        channel_number = ch_info['packetLocation'] + 1  # 1-indexed
                        channels.append({
                            'name': f'ai{channel_number}',
                            'type': 'analog_in',
                            'number': channel_number,
                            'time_channel': 1
                        })

        # Sort channels by type and number
        channels.sort(key=lambda x: (x.get('type', ''), x.get('number', 0)))

        return channels

    def samplerate(self, epochfiles: List[str], channeltype: Any,
                  channel: List[int]) -> List[float]:
        """
        Get the sample rate for specific epoch and channel.

        Args:
            epochfiles: List of file paths for this epoch
            channeltype: Type of channel (not used - same for all)
            channel: List of channel numbers (not used - same for all)

        Returns:
            List of sample rates (same value repeated)

        Note:
            Sampling rate is the same for all channels in SpikeGadgets.
        """
        filename = self.filenamefromepochfiles(epochfiles)
        fileconfig, _ = self._read_spikegadgets_config(filename)

        if fileconfig and 'samplingRate' in fileconfig:
            sr_val = float(fileconfig['samplingRate'])
        else:
            sr_val = 30000.0  # Default

        # Return same sample rate for all channels
        return [sr_val] * len(channel)

    def t0_t1(self, epochfiles: List[str]) -> List[List[float]]:
        """
        Return the t0_t1 (beginning and end) epoch times for an epoch.

        Args:
            epochfiles: List of file paths for this epoch

        Returns:
            List containing [t0, t1] pair in seconds
        """
        filename = self.filenamefromepochfiles(epochfiles)
        fileconfig, _ = self._read_spikegadgets_config(filename)

        if not fileconfig:
            return [[0, 0]]

        # Get file parameters
        header_size_bytes = int(fileconfig.get('headerSize', 0)) * 2  # int16 = 2 bytes
        num_channels = int(fileconfig.get('numChannels', 0))
        channel_size_bytes = num_channels * 2  # int16 = 2 bytes
        block_size_bytes = header_size_bytes + 2 + channel_size_bytes

        # Get file size
        file_size = os.path.getsize(filename)

        # Calculate number of data blocks
        num_data_blocks = (file_size - header_size_bytes) / block_size_bytes

        # Calculate total time
        total_samples = num_data_blocks
        sampling_rate = float(fileconfig.get('samplingRate', 30000))
        total_time = (total_samples - 1) / sampling_rate

        t0 = 0
        t1 = total_time

        return [[t0, t1]]

    def getepochprobemap(self, epochmapfilename: str, epochfiles: List[str]) -> List[Any]:
        """
        Get epoch probe map with probe information (name, reference, n-trode, channels).

        Args:
            epochmapfilename: Epoch map filename (not used for SpikeGadgets)
            epochfiles: List of file paths for this epoch

        Returns:
            List of epoch probe map objects
        """
        filename = self.filenamefromepochfiles(epochfiles)
        fileconfig, _ = self._read_spikegadgets_config(filename)

        if not fileconfig or 'nTrodes' not in fileconfig:
            return []

        epochprobemap = []
        nTrodes = fileconfig['nTrodes']

        for ntrode in nTrodes:
            name = f"Tetrode{ntrode['id']}"
            reference = 1
            probe_type = 'n-trode'
            channels = []

            # Get channels for this nTrode
            if 'channelInfo' in ntrode:
                for ch_info in ntrode['channelInfo']:
                    channels.append(ch_info['packetLocation'] + 1)

            # Create device string
            from ...system.daqsystemstring import DAQSystemString
            devicestring_obj = DAQSystemString(
                'SpikeGadgets',
                ['ai'] * len(channels),
                channels
            )
            devicestring = devicestring_obj.devicestring()

            # Create epoch probe map
            # TODO: Subject needs to be specified somehow
            from ....epoch import EpochProbeMapDAQSystem
            obj = EpochProbeMapDAQSystem(
                name,
                reference,
                probe_type,
                devicestring,
                'subject@lab.org'  # Placeholder
            )

            epochprobemap.append(obj)

        return epochprobemap

    def readchannels_epochsamples(self, channeltype: Any, channels: List[int],
                                  epochfiles: List[str], s0: int, s1: int) -> np.ndarray:
        """
        Read the data based on specified channels.

        Args:
            channeltype: Type of channel to read (string or list of strings)
            channels: List of channel numbers to read (1-indexed)
            epochfiles: List of file paths for this epoch
            s0: Start sample (1-indexed)
            s1: End sample (1-indexed)

        Returns:
            Data array where each column contains data from an individual channel
        """
        filename = self.filenamefromepochfiles(epochfiles)
        header, _ = self._read_spikegadgets_config(filename)

        if not header:
            return np.array([])

        sr = self.samplerate(epochfiles, channeltype, channels)
        detailed_channels = self._get_detailed_channels(epochfiles)

        data = []

        # Ensure channeltype is a list
        if not isinstance(channeltype, list):
            channeltype = [channeltype]

        ct = channeltype[0]

        # Read based on channel type
        from ...system.mfdaq import MFDAQSystem
        mfdaq_type = MFDAQSystem.mfdaq_type(ct)

        if mfdaq_type in ['analog_in', 'analog_out']:
            # Read analog nTrode channels
            # Channels are 0-indexed in file
            data = self._read_spikegadgets_trode_channels(
                filename,
                header['numChannels'],
                [ch - 1 for ch in channels],  # Convert to 0-indexed
                sr[0],
                header['headerSize'],
                s0,
                s1
            )

        elif ct in ['auxiliary', 'aux']:
            # Read auxiliary channels
            byteandbit = []
            for ch in channels:
                # Find byte location in detailed channels
                for det_ch in detailed_channels:
                    if det_ch.get('type') == 'auxiliary' and det_ch.get('number') == ch:
                        byteandbit.append(int(det_ch['startbyte']))
                        break

            data = self._read_spikegadgets_analog_channels(
                filename,
                header['numChannels'],
                byteandbit,
                sr[0],
                header['headerSize'],
                s0,
                s1
            )

        elif ct in ['digital_in', 'digital_out']:
            # Read digital channels
            byteandbit = []
            for ch in channels:
                # Find byte and bit location in detailed channels
                for det_ch in detailed_channels:
                    if det_ch.get('type') == ct and det_ch.get('number') == ch:
                        byteandbit.append([
                            int(det_ch['startbyte']),
                            int(det_ch['bit']) + 1
                        ])
                        break

            data = self._read_spikegadgets_digital_channels(
                filename,
                header['numChannels'],
                byteandbit,
                sr[0],
                header['headerSize'],
                s0,
                s1
            )

            if isinstance(data, np.ndarray):
                data = data.T

        return data if isinstance(data, np.ndarray) else np.array(data)

    def filenamefromepochfiles(self, filename_array: List[str]) -> str:
        """
        Extract the .rec filename from epoch files.

        Args:
            filename_array: List of file paths

        Returns:
            Path to .rec file

        Raises:
            ValueError: If no .rec file found or multiple .rec files
        """
        rec_pattern = r'.*\.rec$'
        matching_files = []

        for filepath in filename_array:
            if re.search(rec_pattern, filepath, re.IGNORECASE):
                matching_files.append(filepath)

        if len(matching_files) > 1:
            raise ValueError('Need only 1 .rec file per epoch.')
        elif len(matching_files) == 0:
            raise ValueError('Need 1 .rec file per epoch.')

        return matching_files[0]

    # Private helper methods

    def _parse_channel_number(self, name: str, prefix: str) -> int:
        """
        Parse channel number from name.

        Args:
            name: Channel name (e.g., 'Ain5', 'Din3')
            prefix: Prefix to remove (e.g., 'Ain', 'Din')

        Returns:
            Channel number
        """
        try:
            return int(name.replace(prefix, ''))
        except ValueError:
            return 0

    def _get_detailed_channels(self, epochfiles: List[str]) -> List[Dict[str, Any]]:
        """
        Get detailed channel information (internal method).

        Args:
            epochfiles: List of file paths for this epoch

        Returns:
            List of channel dictionaries with extra details

        Note:
            This is similar to getchannelsepoch but keeps extra fields
            like 'startbyte', 'bit', 'number' for internal use.
        """
        filename = self.filenamefromepochfiles(epochfiles)
        fileconfig, channels_raw = self._read_spikegadgets_config(filename)

        channels = []

        # Process channels with extra details
        for ch_raw in channels_raw:
            ch_name = ch_raw.get('name', '')
            ch_dict = ch_raw.copy()

            # Parse channel type and number
            if ch_name.startswith('Ain'):
                ch_dict['type'] = 'auxiliary'
                ch_dict['number'] = self._parse_channel_number(ch_name, 'Ain')
            elif ch_name.startswith('Aout'):
                ch_dict['type'] = 'auxiliary'
                ch_dict['number'] = self._parse_channel_number(ch_name, 'Aout')
            elif ch_name.startswith('Din'):
                ch_dict['type'] = 'digital_in'
                ch_dict['number'] = self._parse_channel_number(ch_name, 'Din')
            elif ch_name.startswith('Dout'):
                ch_dict['type'] = 'digital_out'
                ch_dict['number'] = self._parse_channel_number(ch_name, 'Dout')
            elif ch_name.startswith('MCU_Din'):
                ch_dict['type'] = 'digital_in'
                ch_dict['number'] = self._parse_channel_number(ch_name, 'MCU_Din') + 32

            channels.append(ch_dict)

        return channels

    # Placeholders for external file reading functions

    def _read_spikegadgets_config(self, filename: str) -> Tuple[Optional[Dict], List[Dict]]:
        """
        Read SpikeGadgets .rec file configuration.

        Args:
            filename: Path to .rec file

        Returns:
            Tuple of (fileconfig, channels) where:
            - fileconfig: Configuration dictionary
            - channels: List of channel dictionaries

        Note:
            This is a placeholder for the external read_SpikeGadgets_config function.
            TODO: Implement or import from vhlab-thirdparty-python when available.
        """
        # TODO: Implement actual SpikeGadgets config reading
        return {
            'samplingRate': '30000',
            'numChannels': '0',
            'headerSize': '0',
            'nTrodes': []
        }, []

    def _read_spikegadgets_trode_channels(self, filename: str, num_channels: Any,
                                         channels: List[int], sampling_rate: float,
                                         header_size: Any, s0: int, s1: int) -> np.ndarray:
        """
        Read SpikeGadgets nTrode channels.

        Args:
            filename: Path to .rec file
            num_channels: Number of channels
            channels: List of channel numbers (0-indexed)
            sampling_rate: Sampling rate
            header_size: Header size
            s0: Start sample
            s1: End sample

        Returns:
            Data array

        Note:
            This is a placeholder for the external read_SpikeGadgets_trodeChannels function.
            TODO: Implement when available.
        """
        # TODO: Implement actual trode channel reading
        return np.array([])

    def _read_spikegadgets_analog_channels(self, filename: str, num_channels: Any,
                                          byteandbit: List[int], sampling_rate: float,
                                          header_size: Any, s0: int, s1: int) -> np.ndarray:
        """
        Read SpikeGadgets analog channels.

        Args:
            filename: Path to .rec file
            num_channels: Number of channels
            byteandbit: List of byte locations
            sampling_rate: Sampling rate
            header_size: Header size
            s0: Start sample
            s1: End sample

        Returns:
            Data array

        Note:
            This is a placeholder for the external read_SpikeGadgets_analogChannels function.
            TODO: Implement when available.
        """
        # TODO: Implement actual analog channel reading
        return np.array([])

    def _read_spikegadgets_digital_channels(self, filename: str, num_channels: Any,
                                           byteandbit: List[List[int]], sampling_rate: float,
                                           header_size: Any, s0: int, s1: int) -> np.ndarray:
        """
        Read SpikeGadgets digital channels.

        Args:
            filename: Path to .rec file
            num_channels: Number of channels
            byteandbit: List of [byte, bit] pairs
            sampling_rate: Sampling rate
            header_size: Header size
            s0: Start sample
            s1: End sample

        Returns:
            Data array

        Note:
            This is a placeholder for the external read_SpikeGadgets_digitalChannels function.
            TODO: Implement when available.
        """
        # TODO: Implement actual digital channel reading
        return np.array([])

    def __repr__(self) -> str:
        """String representation."""
        return f"SpikeGadgets(name='{self.name if hasattr(self, 'name') else ''}')"

    def __str__(self) -> str:
        """String representation."""
        return self.__repr__()
